pkg/profiles/tap.go

```diff
@@ -0,0 +1,136 @@
+package profiles
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/linkerd/linkerd2/controller/api/util"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// RenderTapOutputProfile performs a tap on the desired resource and generates
+// a service profile with routes pre-populated from the tap data
+// Only inbound tap traffic is considered.
+func RenderTapOutputProfile(client pb.ApiClient, tapResource, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int, w io.Writer) error {
+	requestParams := util.TapRequestParams{
+		Resource:  tapResource,
+		Namespace: namespace,
+	}
+	log.Debugf("Running `linkerd tap %s --namespace %s`", tapResource, namespace)
+
+	req, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		return err
+	}
+
+	profile, err := tapToServiceProfile(client, req, namespace, name, controlPlaneNamespace, tapDuration, routeLimit)
+	if err != nil {
+		return err
+	}
+
+	output, err := yaml.Marshal(profile)
+	if err != nil {
+		return fmt.Errorf("Error writing Service Profile: %s", err)
+	}
+	w.Write(output)
+	return nil
+}
+
+func tapToServiceProfile(client pb.ApiClient, tapReq *pb.TapByResourceRequest, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int) (sp.ServiceProfile, error) {
+	profile := sp.ServiceProfile{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace),
+			Namespace: controlPlaneNamespace,
+		},
+		TypeMeta: ServiceProfileMeta,
+	}
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(tapDuration))
+	defer cancel()
+
+	tapClient, err := client.TapByResource(ctx, tapReq)
+	if err != nil {
+		return profile, err
+	}
+
+	routes := routeSpecFromTap(tapClient, routeLimit)
+
+	profile.Spec.Routes = routes
+
+	return profile, nil
+}
+
+func routeSpecFromTap(tapClient pb.Api_TapByResourceClient, routeLimit int) []*sp.RouteSpec {
+	routes := make([]*sp.RouteSpec, 0)
+	routesMap := make(map[string]*sp.RouteSpec)
+
+	for {
+		log.Debug("Waiting for data...")
+		event, err := tapClient.Recv()
+
+		if err != nil {
+			// expected errors when hitting the tapDuration deadline
+			if err != io.EOF &&
+				!strings.HasSuffix(err.Error(), context.DeadlineExceeded.Error()) &&
```

Comment:
Is there a way to check the type or value of the error object instead of comparing the string message?

Reply:
This unfortunate string check was my suggestion. The error object that contains this string is private:
https://github.com/golang/net/blob/351d144fa1fc0bd934e2408202be0c29f25e35a0/http2/transport.go#L2008

Reply:
:sob:

Reply:
😭